samcli/lib/test_runner/test_runner_template_generator.py

```diff
@@ -1,276 +1,215 @@
+                        action_list=action_list, arn=execute_api_arn
+                    )
+                else:
+                    return jinja2.Template(source=new_statement_template, keep_trailing_newline=True).render(
+                        action_list=action_list, arn=resource_arn
+                    )
+
+        # If the supplied ARN is for a resource without any default actions supported
+        LOG.debug("Found no match for ARN `%s` to any IAM actions.", resource_arn)
+        return jinja2.Template(source=new_statement_template, keep_trailing_newline=True).render(
+            action_list=[], arn=resource_arn
+        )
 
-    image_uri : str
-        The URI of the Image to be used by the Test Runner Fargate task definition.
+    def _get_default_bucket_name(self) -> str:
+        """
+        Returns a guaranteed unique default S3 Bucket name
+        """
+        return "test-runner-bucket" + str(uuid.uuid4())
 
-    tag_filters : dict
-        The tag filters to restrict generating default IAM action permissions to those resources with the given tags.
+    def generate_test_runner_template_string(self, image_uri: str) -> Union[dict, None]:
+        """
+        Renders a base jinja template to create the Test Runner Stack.
 
-        NOTE: https://docs.aws.amazon.com/cli/latest/reference/resourcegroupstaggingapi/get-resources.html#options
+        Parameters
+        ---------
 
-    Returns
-    -------
-    dict
-        The Test Runner CloudFormation template in the form of a YAML string.
+        image_uri : str
+            The URI of the Image to be used by the Test Runner Fargate task definition.
 
-    Raises
-    ------
-    KeyError:
-        If the specified tag does not match any resources.
+        Returns
+        -------
+        dict
+            The Test Runner CloudFormation template in the form of a YAML string.
 
-    jinja2.exceptions.TemplateError:
-        If the IAM Statement template render fails, or if the overall template render fails.
+        Raises
+        ------
+        TestRunnerTemplateGenerationException
+            If the template generation process fails
+        """
 
-    botocore.ClientError
-        If the `get_resources` call fails.
-    """
-    if tag_filters:
+        default_bucket_name = self._get_default_bucket_name()
+        try:
+            jinja_base_template = self._get_jinja_template_string()
+        except FileNotFoundError as jinja_template_not_found_ex:
+            raise TestRunnerTemplateGenerationException(
+                f"Failed to open jinja template, File Not Found."
+            ) from jinja_template_not_found_ex
+
+        if self.resource_arn_list:
+
+            statements_list = []
+            for arn in self.resource_arn_list:
+                try:
+                    statement = self._create_iam_statement_string(arn)
+                except jinja2.exceptions.TemplateError as template_error:
```

Comment:
I know this is a refactor but I assume we added `jinja` to our dependencies?

Any reason we need to use jinja here instead of just string manipulation? We have tended to do string manipulation in the past and output a template that way (reduces the need for additional dependencies).

Reply:
[Jinja is a dependency of Flask](https://flask.palletsprojects.com/en/2.1.x/installation/#dependencies), so it was already in the SAMCLI, I didn't add it.

Reply:
I know that but you need to be adding it to our dependencies. For example, since we directly use Click, Click is a dependency directly. This allows us to pin to specific versions we need from the CLI. Let's say Flask removes Jinja, now you just impacted SAM CLI. Direct dependencies (one you use/import) MUST be defined in our requirements file.

Still curious why `jinja` is needed as a dependency (directly).

Reply:
Right, I see.

I don't think Jinja is needed, I just thought it would be more convenient than direct string manipulation, but it makes sense that it is not worth adding another dependency